pyquil/quil.py

```diff
@@ -348,29 +374,38 @@ def reset(self, qubit_index: Optional[int] = None) -> "Program":
         """
         return self.inst(RESET(qubit_index))
 
-    def measure_all(self, *qubit_reg_pairs):
+    def measure_all(
+        self, *qubit_reg_pairs: Tuple[QubitDesignator, Optional[MemoryReferenceDesignator]]
+    ) -> "Program":
         """
         Measures many qubits into their specified classical bits, in the order
         they were entered. If no qubit/register pairs are provided, measure all qubits present in
         the program into classical addresses of the same index.
 
-        :param Tuple qubit_reg_pairs: Tuples of qubit indices paired with classical bits.
+        :param qubit_reg_pairs: Tuples of qubit indices paired with classical bits.
         :return: The Quil Program with the appropriate measure instructions appended, e.g.
 
         .. code::
 
                   MEASURE 0 [1]
                   MEASURE 1 [2]
                   MEASURE 2 [3]
-
-        :rtype: Program
         """
         if qubit_reg_pairs == ():
-            qubit_inds = self.get_qubits(indices=True)
-            if len(qubit_inds) == 0:
+            qubits = self.get_qubits(indices=True)
+            if len(qubits) == 0:
                 return self
+            if any(isinstance(q, QubitPlaceholder) for q in qubits):
+                raise ValueError(
+                    "Attempted to call measure_all on a Program that contains QubitPlaceholders. "
+                    "You must either provide the qubit_reg_pairs argument to describe how to map "
+                    "these QubitPlaceholders to memory registers, or else first call "
+                    "pyquil.quil.address_qubits to instantiate the QubitPlaceholders."
+                )
+            # Help mypy determine that qubits does not contain any QubitPlaceholders.
+            qubit_inds = cast(List[int], qubits)
             ro = self.declare("ro", "BIT", max(qubit_inds) + 1)
-            for qi in qubit_inds:
+            for qi in qubits:
```

Comment:
This change was not part of the diff I posted. You still want to iterate over `qubit_inds` here.

Reply:
Oh, right, sorry, amended.